core/trino-main/src/main/java/io/trino/operator/exchange/UniformPartitionRebalancer.java

```diff
@@ -0,0 +1,430 @@
+                writerPhysicalWrittenBytesSinceLastRebalance[writerId] = physicalWrittenBytesSinceLastRebalance;
+                writerPriorities[writerId] = physicalWrittenBytesSinceLastRebalance;
+            }
+
+            writerRowCountSinceLastRebalance = new long[numberOfWriters];
+            writerMaxPartitions = new ArrayList<>(numberOfWriters);
+            for (int writerId = 0; writerId < numberOfWriters; writerId++) {
+                writerMaxPartitions.add(new IndexedPriorityQueue<>());
+            }
+
+            partitionRowCounts.forEach((serializedKey, rowCount) -> {
+                WriterPartitionId writerPartitionId = WriterPartitionId.deserialize(serializedKey);
+                writerRowCountSinceLastRebalance[writerPartitionId.writerId] += rowCount;
+                writerMaxPartitions
+                        .get(writerPartitionId.writerId)
+                        .addOrUpdate(new PartitionIdWithRowCount(writerPartitionId.partitionId, rowCount), rowCount);
+            });
+        }
+
+        private boolean rebalancePartition(WriterId from, WriterId to)
+        {
+            IndexedPriorityQueue<PartitionIdWithRowCount> maxPartitions = writerMaxPartitions.get(from.id);
+
+            while (true) {
+                // Find the partition with maximum written bytes since last rebalance
+                PartitionIdWithRowCount partitionToRebalance = maxPartitions.poll();
+
+                if (partitionToRebalance == null) {
+                    return false;
+                }
+
+                PartitionInfo partitionInfo = partitionInfos[partitionToRebalance.id];
+
+                // If partition is already balanced or min skewed writer is already writing to that partition, then skip
+                // this partition and move on to the other partition inside max writer.
+                if (isPartitionRebalanced(partitionToRebalance.id) || partitionInfo.containsWriter(to.id)) {
+                    continue;
+                }
+
+                long estimatedPartitionWrittenBytesSinceLastRebalance = estimatePartitionWrittenBytesSinceLastRebalance(from, partitionToRebalance.rowCount);
+                long estimatedPartitionWrittenBytes = estimatedPartitionWrittenBytesSinceLastRebalance + partitionInfo.getPhysicalWrittenBytesAtLastRebalancePerWriter();
+
+                // Scale the partition when estimated physicalWrittenBytes is greater than writerMinSize.
+                if (partitionInfo.getWriterCount() <= numberOfWriters && estimatedPartitionWrittenBytes >= writerMinSize) {
+                    partitionInfo.addWriter(to.id);
+                    rebalancedPartitions.add(partitionToRebalance.id);
+                    updateWriterPriorities(from, to, estimatedPartitionWrittenBytesSinceLastRebalance);
+                    log.debug(format("Scaled partition (%s) to writer %s with writer count %s", partitionToRebalance.id, to.id, partitionInfo.getWriterCount()));
```

Comment:
use build-in `log` formatting rather than `format`. This way string is not constructed for each debug call.